### ISSUE TITLE
The train-error of ImageNet is too big

### ISSUE THREAD

**Issue by dpnight** [dmlc/cxxnet#176]
I use the ImageNet model to train a image classification classifier but the train-error is too big:
update round 0[1]	train-error:0.664583	test-error:0.671429
update round 1[2]	train-error:0.71875	test-error:0.678571
update round 2[3]	train-error:0.64375	test-error:0.671429
update round 3[4]	train-error:0.63125	test-error:0.671429
update round 4[5]	train-error:0.6625	test-error:0.671429
update round 5[6]	train-error:0.702083	test-error:0.671429
update round 6[7]	train-error:0.652083	test-error:0.671429
update round 7[8]	train-error:0.6625	test-error:0.65
update round 8[9]	train-error:0.677083	test-error:0.671429
update round 9[10]	train-error:0.670833	test-error:0.671429
update round 10[11]	train-error:0.69375	test-error:0.671429

---
the dataset is not ImageNet but some images about dog and cat.
the training dataset file like this:
282	2	dog_171.jpg
341	2	dog_44.jpg
384	2	dog_83.jpg
45	1	cat_137.jpg
249	2	dog_141.jpg

---
the configuration file like this:
\# Configuration for ImageNet  
\# Acknowledgement:
\#  Ref: http://www.cs.toronto.edu/~fritz/absps/imagenet.pdf
\#  The scheduling parameters is adapted from Caffe(http://caffe.berkeleyvision.org/)

data = train
iter = img
  image_list = "/home/wy/ImageNet/NameList.train"
\#  image_bin  = "../../TRAIN.BIN"
  image_root = "/home/wy/ImageNet/all/"
\#  image_mean = "models/image_net_mean.bin"
  rand_crop=1
  rand_mirror=1
iter = threadbuffer
iter = end

eval = test
iter = img
  image_list = "/home/wy/ImageNet/NameList.test"
\#  image_bin = "../../TEST.BIN"
  image_root = "/home/wy/ImageNet/all/"
\#  image_mean = "models/image_net_mean.bin"
\# no random crop and mirror in test
iter = end

\#pred = pred.txt
\#iter = img
\#	image_list = "/home/wy/ImageNet/NameList.test"
\#	image_root = "/home/wy/ImageNet/all/"
\#iter = end

netconfig=start
layer[0->1] = conv:conv1
  kernel_size = 11
  stride = 4
  nchannel = 96
layer[1->2] = relu:relu1
layer[2->3] = max_pooling:pool1
  kernel_size = 3
  stride = 2
layer[3->4] = lrn:lrn1
  local_size = 5
  alpha = 0.0001
  beta = 0.75
  knorm = 1
\###############
layer[4->5] = conv:conv2
  ngroup = 2
  nchannel = 256
  kernel_size = 5
  pad = 2
layer[5->6] = relu:relu2
layer[6->7] = max_pooling:pool2
  kernel_size = 3
  stride = 2
layer[7->8] = lrn:lrn2
  local_size = 5
  alpha = 0.0001
  beta = 0.75
  knorm = 1
\#############
layer[8->9] = conv:conv3
  nchannel = 384
  kernel_size = 3
  pad = 1
layer[9->10]= relu:relu3
layer[10->11] = conv:conv4
  nchannel = 384
  ngroup = 2
  kernel_size = 3
  pad = 1
layer[11->12] = relu:relu4
layer[12->13] = conv:conv5
  nchannel = 256
  ngroup = 2
  kernel_size = 3
  pad = 1
  init_bias = 1.0
layer[13->14] = relu:relu5
layer[14->15] = max_pooling:pool5
  kernel_size = 3
  stride = 2
layer[15->16] = flatten:flatten1
layer[16->17] = fullc:fc6
  nhidden = 4096
  init_sigma = 0.005
  init_bias = 1.0
layer[17->18] = relu:relu6
layer[18->18] = dropout:dropout1
  threshold = 0.5
layer[18->19] = fullc:fc7
  nhidden = 4096
  init_sigma = 0.005
  init_bias = 1.0
layer[19->20] = relu:relu7
layer[20->20] = dropout:dropout2
  threshold = 0.5
layer[20->21] = fullc:fc8
  nhidden = 1000
layer[21->21] = softmax:softmax1
netconfig=end

\# evaluation metric
metric = error
\#metric = rec@1
\#metric = rec@5

max_round = 45
num_round = 45

\# input shape not including batch
input_shape = 3,227,227

batch_size = 40

\# global parameters in any sectiion outside netconfig, and iter
dev = gpu
CUDNN = 0
momentum = 0.9
wmat:lr  = 0.01
wmat:wd  = 0.0005

bias:wd  = 0.000
bias:lr  = 0.02

\# all the learning rate schedule starts with lr
lr:schedule = expdecay
lr:gamma = 0.1
lr:step = 100000

save_model=1
model_dir=models

\#task = pred
#model_in = ./models/0045.model

\# random config
random_type = gaussian
init_sigma = 0.01

\# new line

**Comment by dpnight** [comment]
I don't know how to convert image to bin, so I just use iter = img. 
But I think this is not the reason. 
Can someone help me?

**Comment by antinucleon** [comment]
The main problem is you don't set mean file. Using img iter is extremely slow. Please check [here](https://github.com/dmlc/cxxnet/blob/master/doc/io.md#image-iterator) to generate a image list, then use ```bin/im2bin``` or ```bin/im2rec``` to generate binary format data.

**Comment by dpnight** [comment]
@antinucleon thank you!
I can not find >models/mean_224.bin which is used in >image_mean = "models/mean_224.bin"

**Comment by antinucleon** [comment]
For first problem: you need to create a folder ```models``` then  it will create the mean file automatically. For second problem, you can use pretrained "GoogLeNet". More detains can be found [here](https://github.com/dmlc/cxxnet/blob/master/example/ImageNet/Inception-BN.conf)

**Comment by dpnight** [comment]
@antinucleon 
The prediction file just consists of classID, so I download the the 1000 classification categories [here] http://image-net.org/challenges/LSVRC/2014/browse-synsets) and joint them in a list to see what the classID refers to. 
The result is confused. Do your `1000 classification categories file` just like mine?

---
should all Image files be resized into 224*224?

---
Thanks for your great work and your reply!

**Comment by Carydileo** [comment]
Sent from Yahoo Mail on Android

**Comment by dzhwinter** [comment]
according to my test of im2bin, it failed on binarypage.push(fobj) in line 54 of im2bin.cpp and generate an fix sized binary file of 64MB(of course, the binary pagesize).   I have checked my path and image list format a third time.   Did anyone come across the same problem with me?

**Comment by winstywang** [comment]
@dzhwinter Could you please try the latest im2rec? im2bin has been deprecated. I have just updated the doc of imgrec iterator and im2rec. Could these help: https://github.com/dmlc/cxxnet/blob/master/doc/io.md https://github.com/dmlc/cxxnet/blob/master/example/ImageNet/README.md

**Comment by dzhwinter** [comment]
already Done yet. see my issue #182
